### PR TITLE
Add a function to get memory size of array slice

### LINKED ISSUE #3407: ArrayData`get_slice_memory_size`   or similar
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As part of implementing an Arrow flight client, we have observed that there is a maximum message size for gRPC messages, and thus while encoding the Arrow flight client has a way to limit the size of the messages (see https://github.com/apache/arrow-rs/pull/3391)

However, apparently there is currently no way to get the memory size used by a sliced (aka the array only points to a portion of the overall underlying allocation), which will result in overly conservative re-splitting in the arrow-flight client 

**Describe the solution you'd like**
As suggested by @tustvold  on https://github.com/apache/arrow-rs/pull/3391#discussion_r1058288317

> It is worth noting that this method does not take into account already sliced arrays, we likely need to add a `get_slice_memory_size` or something to `ArrayData` to provide this, it should be relatively straightforward

**Describe alternatives you've considered**
We could leave the users of the client to worry about batch splitting

**Additional context**

### PR DESCRIPTION
**Pull request by askoa**
# Which issue does this PR close?

Closes #3407 

# Rationale for this change
See issue description.

# What changes are included in this PR?
Function to calculate buffer size of array slice.

# Are there any user-facing changes?
User will get a new API that'll calculate the buffer size of array slice.

### CODE REVIEW

File: arrow-data/src/data.rs [hunk 1]
```diff
@@ -463,6 +463,59 @@ impl ArrayData {
...
+        let mut result: usize = 0;
+        let layout = layout(&self.data_type);
+
+        for spec in layout.buffers.iter() {
+            match spec {
+                BufferSpec::FixedWidth { byte_width } => {
+                    let buffer_size = self
+                        .len
+                        .checked_mul(*byte_width)
+                        .expect("integer overflow computing buffer size");
```
Comment (viirya) [inline arrow-data/src/data.rs:476]: Instead of panic, it should be better to return `Err`.

File: arrow-data/src/data.rs [hunk 2]
```diff
@@ -463,6 +463,59 @@ impl ArrayData {
         size
     }
 
+    pub fn get_slice_memory_size(&self) -> Result<usize, ArrowError> {
```
Comment (viirya) [inline arrow-data/src/data.rs:466]: For public api, it is better to add a doc.

**Review by tustvold — APPROVED** [review]
It is worth noting that this may overestimate for sliced ListArray and StructArray, but I'm not sure if we are being smart when writing those yet. Either way I think this is a good start

@viirya may know??

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
For `StructArray`, its layout is empty so this basically computes on all its children. Looks like when slicing `StructArray`, we already did correctly for its child data? Then seems okay.

I'm not sure if there is an issue on `ListArray`. But seems it is same as other types where have one offset buffer + child data?

**Comment by viirya** [comment]
> As for lists, theoretically the IPC writer could find the min offset in the list slice, and then shift everything down - as is done for StringArray, etc... Not sure if we do this yet...

Oh, I see. I remember I did something on buffer slice in IPC writer. I may need to re-look it to see if we do it for list slice.

**Comment by alamb** [comment]
Wow -- thank you @askoa . This is great

**Pull request merged**